<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

<h1 align=center><font size = 5>Assignment: Notebook for Graded Assessment</font></h1>


# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a SQLIte database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. 

Use the links below to read the data files using the Pandas library. 

* Chicago Census Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Public Schools

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Crime Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

**NOTE:** Ensure you use the datasets available on the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


### Store the datasets in database tables

To analyze the data using SQL, it first needs to be loaded into SQLite DB.
We will create three tables in as under:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**


Load the `pandas` and `sqlite3` libraries and establish a connection to `FinalDB.db`


In [9]:
# Import the necessary libraries
import pandas as pd
import sqlite3

# Establish a connection to the SQLite database
# If the database does not exist, it will be created
con = sqlite3.connect('FinalDB.db')

# Create a cursor object to execute SQL commands
cur = con.cursor()

# You should now see that you are connected to the database
print("Connected to FinalDB.db")


Connected to FinalDB.db


Load the SQL magic module


In [11]:
# Load the SQL extension
%load_ext sql

# Connect the SQL magic to your SQLite database
%sql sqlite:///FinalDB.db


Use `Pandas` to load the data available in the links above to dataframes. Use these dataframes to load data on to the database `FinalDB.db` as required tables.


In [13]:
#Load Data into Pandas DataFrames
# URLs of the datasets
census_data_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv"
schools_data_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv"
crime_data_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv"

# Load data into Pandas DataFrames
df_census = pd.read_csv(census_data_url)
df_schools = pd.read_csv(schools_data_url)
df_crime = pd.read_csv(crime_data_url)

# Display the first few rows of each DataFrame to verify the data
print("Census Data:")
print(df_census.head())

print("\nChicago Public Schools Data:")
print(df_schools.head())

print("\nChicago Crime Data:")
print(df_crime.head())


Census Data:
   COMMUNITY_AREA_NUMBER COMMUNITY_AREA_NAME  PERCENT_OF_HOUSING_CROWDED  \
0                    1.0         Rogers Park                         7.7   
1                    2.0          West Ridge                         7.8   
2                    3.0              Uptown                         3.8   
3                    4.0      Lincoln Square                         3.4   
4                    5.0        North Center                         0.3   

   PERCENT_HOUSEHOLDS_BELOW_POVERTY  PERCENT_AGED_16__UNEMPLOYED  \
0                              23.6                          8.7   
1                              17.2                          8.8   
2                              24.0                          8.9   
3                              10.9                          8.2   
4                               7.5                          5.2   

   PERCENT_AGED_25__WITHOUT_HIGH_SCHOOL_DIPLOMA  \
0                                          18.2   
1                  

In [14]:
#Store DataFrames as Tables in SQLite Database
df_census.to_sql('CENSUS_DATA', con, if_exists='replace', index=False)
df_schools.to_sql('CHICAGO_PUBLIC_SCHOOLS', con, if_exists='replace', index=False)
df_crime.to_sql('CHICAGO_CRIME_DATA', con, if_exists='replace', index=False)

print("Data has been loaded into the FinalDB.db database.")



Data has been loaded into the FinalDB.db database.


In [15]:
%sql PRAGMA table_info(CENSUS_DATA);


 * sqlite:///FinalDB.db
Done.


[(0, 'COMMUNITY_AREA_NUMBER', 'REAL', 0, None, 0),
 (1, 'COMMUNITY_AREA_NAME', 'TEXT', 0, None, 0),
 (2, 'PERCENT_OF_HOUSING_CROWDED', 'REAL', 0, None, 0),
 (3, 'PERCENT_HOUSEHOLDS_BELOW_POVERTY', 'REAL', 0, None, 0),
 (4, 'PERCENT_AGED_16__UNEMPLOYED', 'REAL', 0, None, 0),
 (5, 'PERCENT_AGED_25__WITHOUT_HIGH_SCHOOL_DIPLOMA', 'REAL', 0, None, 0),
 (6, 'PERCENT_AGED_UNDER_18_OR_OVER_64', 'REAL', 0, None, 0),
 (7, 'PER_CAPITA_INCOME', 'INTEGER', 0, None, 0),
 (8, 'HARDSHIP_INDEX', 'REAL', 0, None, 0)]

Establish a connection between SQL magic module and the database `FinalDB.db`


In [17]:
%sql PRAGMA table_info(CHICAGO_PUBLIC_SCHOOLS);

 * sqlite:///FinalDB.db
Done.


[(0, 'School_ID', 'INTEGER', 0, None, 0),
 (1, 'NAME_OF_SCHOOL', 'TEXT', 0, None, 0),
 (2, 'Elementary, Middle, or High School', 'TEXT', 0, None, 0),
 (3, 'Street_Address', 'TEXT', 0, None, 0),
 (4, 'City', 'TEXT', 0, None, 0),
 (5, 'State', 'TEXT', 0, None, 0),
 (6, 'ZIP_Code', 'INTEGER', 0, None, 0),
 (7, 'Phone_Number', 'TEXT', 0, None, 0),
 (8, 'Link', 'TEXT', 0, None, 0),
 (9, 'Network_Manager', 'TEXT', 0, None, 0),
 (10, 'Collaborative_Name', 'TEXT', 0, None, 0),
 (11, 'Adequate_Yearly_Progress_Made_', 'TEXT', 0, None, 0),
 (12, 'Track_Schedule', 'TEXT', 0, None, 0),
 (13, 'CPS_Performance_Policy_Status', 'TEXT', 0, None, 0),
 (14, 'CPS_Performance_Policy_Level', 'TEXT', 0, None, 0),
 (15, 'HEALTHY_SCHOOL_CERTIFIED', 'TEXT', 0, None, 0),
 (16, 'Safety_Icon', 'TEXT', 0, None, 0),
 (17, 'SAFETY_SCORE', 'REAL', 0, None, 0),
 (18, 'Family_Involvement_Icon', 'TEXT', 0, None, 0),
 (19, 'Family_Involvement_Score', 'TEXT', 0, None, 0),
 (20, 'Environment_Icon', 'TEXT', 0, None, 0),
 (21, 'Environment_Score', 'REAL', 0, None, 0),
 (22, 'Instruction_Icon', 'TEXT', 0, None, 0),
 (23, 'Instruction_Score', 'REAL', 0, None, 0),
 (24, 'Leaders_Icon', 'TEXT', 0, None, 0),
 (25, 'Leaders_Score', 'TEXT', 0, None, 0),
 (26, 'Teachers_Icon', 'TEXT', 0, None, 0),
 (27, 'Teachers_Score', 'TEXT', 0, None, 0),
 (28, 'Parent_Engagement_Icon', 'TEXT', 0, None, 0),
 (29, 'Parent_Engagement_Score', 'TEXT', 0, None, 0),
 (30, 'Parent_Environment_Icon', 'TEXT', 0, None, 0),
 (31, 'Parent_Environment_Score', 'TEXT', 0, None, 0),
 (32, 'AVERAGE_STUDENT_ATTENDANCE', 'TEXT', 0, None, 0),
 (33, 'Rate_of_Misconducts__per_100_students_', 'REAL', 0, None, 0),
 (34, 'Average_Teacher_Attendance', 'TEXT', 0, None, 0),
 (35, 'Individualized_Education_Program_Compliance_Rate', 'TEXT', 0, None, 0),
 (36, 'Pk_2_Literacy__', 'TEXT', 0, None, 0),
 (37, 'Pk_2_Math__', 'TEXT', 0, None, 0),
 (38, 'Gr3_5_Grade_Level_Math__', 'TEXT', 0, None, 0),
 (39, 'Gr3_5_Grade_Level_Read__', 'TEXT', 0, None, 0),
 (40, 'Gr3_5_Keep_Pace_Read__', 'TEXT', 0, None, 0),
 (41, 'Gr3_5_Keep_Pace_Math__', 'TEXT', 0, None, 0),
 (42, 'Gr6_8_Grade_Level_Math__', 'TEXT', 0, None, 0),
 (43, 'Gr6_8_Grade_Level_Read__', 'TEXT', 0, None, 0),
 (44, 'Gr6_8_Keep_Pace_Math_', 'TEXT', 0, None, 0),
 (45, 'Gr6_8_Keep_Pace_Read__', 'TEXT', 0, None, 0),
 (46, 'Gr_8_Explore_Math__', 'TEXT', 0, None, 0),
 (47, 'Gr_8_Explore_Read__', 'TEXT', 0, None, 0),
 (48, 'ISAT_Exceeding_Math__', 'REAL', 0, None, 0),
 (49, 'ISAT_Exceeding_Reading__', 'REAL', 0, None, 0),
 (50, 'ISAT_Value_Add_Math', 'REAL', 0, None, 0),
 (51, 'ISAT_Value_Add_Read', 'REAL', 0, None, 0),
 (52, 'ISAT_Value_Add_Color_Math', 'TEXT', 0, None, 0),
 (53, 'ISAT_Value_Add_Color_Read', 'TEXT', 0, None, 0),
 (54, 'Students_Taking__Algebra__', 'TEXT', 0, None, 0),
 (55, 'Students_Passing__Algebra__', 'TEXT', 0, None, 0),
 (56, '9th Grade EXPLORE (2009)', 'TEXT', 0, None, 0),
 (57, '9th Grade EXPLORE (2010)', 'TEXT', 0, None, 0),
 (58, '10th Grade PLAN (2009)', 'TEXT', 0, None, 0),
 (59, '10th Grade PLAN (2010)', 'TEXT', 0, None, 0),
 (60, 'Net_Change_EXPLORE_and_PLAN', 'TEXT', 0, None, 0),
 (61, '11th Grade Average ACT (2011)', 'TEXT', 0, None, 0),
 (62, 'Net_Change_PLAN_and_ACT', 'TEXT', 0, None, 0),
 (63, 'College_Eligibility__', 'TEXT', 0, None, 0),
 (64, 'Graduation_Rate__', 'TEXT', 0, None, 0),
 (65, 'College_Enrollment_Rate__', 'TEXT', 0, None, 0),
 (66, 'COLLEGE_ENROLLMENT', 'INTEGER', 0, None, 0),
 (67, 'General_Services_Route', 'INTEGER', 0, None, 0),
 (68, 'Freshman_on_Track_Rate__', 'TEXT', 0, None, 0),
 (69, 'X_COORDINATE', 'REAL', 0, None, 0),
 (70, 'Y_COORDINATE', 'REAL', 0, None, 0),
 (71, 'Latitude', 'REAL', 0, None, 0),
 (72, 'Longitude', 'REAL', 0, None, 0),
 (73, 'COMMUNITY_AREA_NUMBER', 'INTEGER', 0, None, 0),
 (74, 'COMMUNITY_AREA_NAME', 'TEXT', 0, None, 0),
 (75, 'Ward', 'INTEGER', 0, None, 0),
 (76, 'Police_District', 'INTEGER', 0, None, 0),
 (77, 'Location', 'TEXT', 0, None, 0)]

In [18]:
%sql PRAGMA table_info(CHICAGO_CRIME_DATA);

 * sqlite:///FinalDB.db
Done.


[(0, 'ID', 'INTEGER', 0, None, 0),
 (1, 'CASE_NUMBER', 'TEXT', 0, None, 0),
 (2, 'DATE', 'TEXT', 0, None, 0),
 (3, 'BLOCK', 'TEXT', 0, None, 0),
 (4, 'IUCR', 'TEXT', 0, None, 0),
 (5, 'PRIMARY_TYPE', 'TEXT', 0, None, 0),
 (6, 'DESCRIPTION', 'TEXT', 0, None, 0),
 (7, 'LOCATION_DESCRIPTION', 'TEXT', 0, None, 0),
 (8, 'ARREST', 'INTEGER', 0, None, 0),
 (9, 'DOMESTIC', 'INTEGER', 0, None, 0),
 (10, 'BEAT', 'INTEGER', 0, None, 0),
 (11, 'DISTRICT', 'INTEGER', 0, None, 0),
 (12, 'WARD', 'REAL', 0, None, 0),
 (13, 'COMMUNITY_AREA_NUMBER', 'REAL', 0, None, 0),
 (14, 'FBICODE', 'TEXT', 0, None, 0),
 (15, 'X_COORDINATE', 'REAL', 0, None, 0),
 (16, 'Y_COORDINATE', 'REAL', 0, None, 0),
 (17, 'YEAR', 'INTEGER', 0, None, 0),
 (18, 'LATITUDE', 'REAL', 0, None, 0),
 (19, 'LONGITUDE', 'REAL', 0, None, 0),
 (20, 'LOCATION', 'TEXT', 0, None, 0)]

In [19]:
%load_ext sql
%sql sqlite:///FinalDB.db
%sql SELECT name FROM sqlite_master WHERE type='table';


The sql extension is already loaded. To reload it, use:
  %reload_ext sql
 * sqlite:///FinalDB.db
Done.


[('CENSUS_DATA',), ('CHICAGO_PUBLIC_SCHOOLS',), ('CHICAGO_CRIME_DATA',)]

In [20]:
import pandas as pd
import sqlite3

# Establish a connection to the database
con = sqlite3.connect("FinalDB.db")

# URL of the CSV file
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv'

# Read the CSV file into a Pandas DataFrame
df = pd.read_csv(url)

# Store the DataFrame in a table named 'CENSUS_DATA'
df.to_sql('CENSUS_DATA', con, if_exists='replace', index=False)

# Verify by selecting the first 5 rows
%sql SELECT * FROM CENSUS_DATA LIMIT 5


 * sqlite:///FinalDB.db
Done.


[(1.0, 'Rogers Park', 7.7, 23.6, 8.7, 18.2, 27.5, 23939, 39.0),
 (2.0, 'West Ridge', 7.8, 17.2, 8.8, 20.8, 38.5, 23040, 46.0),
 (3.0, 'Uptown', 3.8, 24.0, 8.9, 11.8, 22.2, 35787, 20.0),
 (4.0, 'Lincoln Square', 3.4, 10.9, 8.2, 13.4, 25.5, 37524, 17.0),
 (5.0, 'North Center', 0.3, 7.5, 5.2, 4.5, 26.2, 57123, 6.0)]

In [21]:
import pandas as pd
import sqlite3

# Establish a connection to the database
con = sqlite3.connect("FinalDB.db")

# URL of the CSV file
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv'

# Read the CSV file into a Pandas DataFrame
df = pd.read_csv(url)

# Store the DataFrame in a table named 'CENSUS_DATA'
df.to_sql('CHICAGO_PUBLIC_SCHOOLS', con, if_exists='replace', index=False)

# Verify by selecting the first 5 rows
%sql SELECT * FROM CHICAGO_PUBLIC_SCHOOLS LIMIT 5


 * sqlite:///FinalDB.db
Done.


[(610038, 'Abraham Lincoln Elementary School', 'ES', '615 W Kemper Pl', 'Chicago', 'IL', 60614, '(773) 534-5720', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610038.pdf', 'Fullerton Elementary Network', 'NORTH-NORTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 1', 'Yes', 'Very Strong', 99.0, 'Very Strong', '99', 'Strong', 74.0, 'Strong', 66.0, 'Weak', '65', 'Strong', '70', 'Strong', '56', 'Average', '47', '96.00%', 2.0, '96.40%', '95.80%', '80.1', '43.3', '89.6', '84.9', '60.7', '62.6', '81.9', '85.2', '52', '62.4', '66.3', '77.9', 69.7, 64.4, 0.2, 0.9, 'Yellow', 'Green', '67.1', '54.5', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 813, 33, 'NDA', 1171699.458, 1915829.428, 41.92449696, -87.64452163, 7, 'LINCOLN PARK', 43, 18, '(41.92449696, -87.64452163)'),
 (610281, 'Adam Clayton Powell Paideia Community Academy Elementary School', 'ES', '7511 S South Shore Dr', 'Chicago', 'IL', 60649, '(773) 535-6650', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610281.pdf', 'Skyway Elementary Network', 'SOUTH SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 54.0, 'Strong', '66', 'Strong', 74.0, 'Very Strong', 84.0, 'Weak', '63', 'Strong', '76', 'Weak', '46', 'Average', '50', '95.60%', 15.7, '95.30%', '100.00%', '62.4', '51.7', '21.9', '15.1', '29', '42.8', '38.5', '27.4', '44.8', '42.7', '14.1', '34.4', 16.8, 16.5, 0.7, 1.4, 'Green', 'Green', '17.2', '27.3', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 521, 46, 'NDA', 1196129.985, 1856209.466, 41.76032435, -87.55673627, 43, 'SOUTH SHORE', 7, 4, '(41.76032435, -87.55673627)'),
 (610185, 'Adlai E Stevenson Elementary School', 'ES', '8010 S Kostner Ave', 'Chicago', 'IL', 60652, '(773) 535-2280', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610185.pdf', 'Midway Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 2', 'No', 'Strong', 61.0, 'NDA', 'NDA', 'Average', 50.0, 'Weak', 36.0, 'Weak', 'NDA', 'NDA', 'NDA', 'Average', '47', 'Weak', '41', '95.70%', 2.3, '94.70%', '98.30%', '53.7', '26.6', '38.3', '34.7', '43.7', '57.3', '48.8', '39.2', '46.8', '44', '7.5', '21.9', 18.3, 15.5, -0.9, -1.0, 'Red', 'Red', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 1324, 44, 'NDA', 1148427.165, 1851012.215, 41.74711093, -87.73170248, 70, 'ASHBURN', 13, 8, '(41.74711093, -87.73170248)'),
 (609993, 'Agustin Lara Elementary Academy', 'ES', '4619 S Wolcott Ave', 'Chicago', 'IL', 60609, '(773) 535-4389', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_609993.pdf', 'Pershing Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 56.0, 'Average', '44', 'Average', 45.0, 'Weak', 37.0, 'Weak', '65', 'Average', '48', 'Average', '53', 'Strong', '58', '95.50%', 10.4, '95.80%', '100.00%', '76.9', 'NDA', '26', '24.7', '61.8', '49.7', '39.2', '27.2', '69.7', '60.6', '9.1', '18.2', 11.1, 9.6, 0.9, 2.4, 'Green', 'Green', '42.9', '25', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 556, 42, 'NDA', 1164504.29, 1873959.199, 41.8097569, -87.6721446, 61, 'NEW CITY', 20, 9, '(41.8097569, -87.6721446)'),
 (610513, 'Air Force Academy High School', 'HS', '3630 S Wells St', 'Chicago', 'IL', 60609, '(773) 535-1590', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610513.pdf', 'Southwest Side High School Network', 'SOUTHWEST SIDE COLLABORATIVE', 'NDA', 'Standard', 'Not on Probation', 'Not Enough Data', 'Yes', 'Average', 49.0, 'Strong', '60', 'Strong', 60.0, 'Average', 55.0, 'Weak', '45', 'Average', '54', 'Average', '53', 'Average', '49', '93.30%', 15.6, '96.90%', '100.00%', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', None, None, None, None, 'NDA', 'NDA', 'NDA', 'NDA', '14.6', '14.8', 'NDA', '16', '1.4', 'NDA', 

In [22]:
import pandas as pd
import sqlite3

# Establish a connection to the database
con = sqlite3.connect("FinalDB.db")

# URL of the CSV file
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv'

# Read the CSV file into a Pandas DataFrame
df = pd.read_csv(url)

# Store the DataFrame in a table named 'CENSUS_DATA'
df.to_sql('CHICAGO_CRIME_DATA', con, if_exists='replace', index=False)

# Verify by selecting the first 5 rows
%sql SELECT * FROM CHICAGO_CRIME_DATA LIMIT 5


 * sqlite:///FinalDB.db
Done.


[(3512276, 'HK587712', '2004-08-28', '047XX S KEDZIE AVE', '890', 'THEFT', 'FROM BUILDING', 'SMALL RETAIL STORE', 0, 0, 911, 9, 14.0, 58.0, '6', 1155838.0, 1873050.0, 2004, 41.8074405, -87.70395585, '(41.8074405, -87.703955849)'),
 (3406613, 'HK456306', '2004-06-26', '009XX N CENTRAL PARK AVE', '820', 'THEFT', '$500 AND UNDER', 'OTHER', 0, 0, 1112, 11, 27.0, 23.0, '6', 1152206.0, 1906127.0, 2004, 41.89827996, -87.71640551, '(41.898279962, -87.716405505)'),
 (8002131, 'HT233595', '2011-04-04', '043XX S WABASH AVE', '820', 'THEFT', '$500 AND UNDER', 'NURSING HOME/RETIREMENT HOME', 0, 0, 221, 2, 3.0, 38.0, '6', 1177436.0, 1876313.0, 2011, 41.81593313, -87.62464213, '(41.815933131, -87.624642127)'),
 (7903289, 'HT133522', '2010-12-30', '083XX S KINGSTON AVE', '840', 'THEFT', 'FINANCIAL ID THEFT: OVER $300', 'RESIDENCE', 0, 0, 423, 4, 7.0, 46.0, '6', 1194622.0, 1850125.0, 2010, 41.74366532, -87.56246276, '(41.743665322, -87.562462756)'),
 (10402076, 'HZ138551', '2016-02-02', '033XX W 66TH ST', '820', 'THEFT', '$500 AND UNDER', 'ALLEY', 0, 0, 831, 8, 15.0, 66.0, '6', 1155240.0, 1860661.0, 2016, 41.7734553, -87.70648047, '(41.773455295, -87.706480471)')]

You can now proceed to the the following questions. Please note that a graded assignment will follow this lab and there will be a question on each of the problems stated below. It can be from the answer you received or the code you write for this problem. Therefore, please keep a note of both your codes as well as the response you generate.


## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.


In [25]:
%sql SELECT COUNT(*) AS total_crimes FROM CHICAGO_CRIME_DATA;


 * sqlite:///FinalDB.db
Done.


[(533,)]

### Problem 2

##### List community area names and numbers with per capita income less than 11000.


In [27]:
%sql SELECT COMMUNITY_AREA_NUMBER, COMMUNITY_AREA_NAME FROM CENSUS_DATA WHERE PER_CAPITA_INCOME < 11000

 * sqlite:///FinalDB.db
Done.


[(26.0, 'West Garfield Park'),
 (30.0, 'South Lawndale'),
 (37.0, 'Fuller Park'),
 (54.0, 'Riverdale')]

### Problem 3

##### List all case numbers for crimes involving minors?(children are not considered minors for the purposes of crime analysis) 


In [29]:
%sql SELECT CASE_NUMBER FROM CHICAGO_CRIME_DATA WHERE DESCRIPTION LIKE '%minor%' OR DESCRIPTION LIKE '%juvenile%';


 * sqlite:///FinalDB.db
Done.


[('HL266884',), ('HK238408',)]

### Problem 4

##### List all kidnapping crimes involving a child?


In [31]:
%sql SELECT * FROM CHICAGO_CRIME_DATA WHERE PRIMARY_TYPE = "KIDNAPPING" AND DESCRIPTION LIKE "%CHILD%"



 * sqlite:///FinalDB.db
(sqlite3.OperationalError) near "%": syntax error
[SQL: SELECT * FROM CHICAGO_CRIME_DATA WHERE PRIMARY_TYPE = KIDNAPPING AND DESCRIPTION LIKE %CHILD%]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


### Problem 5

##### List the kind of crimes that were recorded at schools. (No repetitions)


In [33]:
%sql SELECT DISTINCT PRIMARY_TYPE FROM CHICAGO_CRIME_DATA WHERE LOCATION_DESCRIPTION LIKE '%SCHOOL%'


 * sqlite:///FinalDB.db
Done.


[('BATTERY',),
 ('CRIMINAL DAMAGE',),
 ('NARCOTICS',),
 ('ASSAULT',),
 ('CRIMINAL TRESPASS',),
 ('PUBLIC PEACE VIOLATION',)]

### Problem 6

##### List the type of schools along with the average safety score for each type.


In [35]:
%sql SELECT "Elementary, Middle, or High School" AS School_Type, AVG(SAFETY_SCORE) AS Average_Safety_Score FROM CHICAGO_PUBLIC_SCHOOLS GROUP BY "Elementary, Middle, or High School"

 * sqlite:///FinalDB.db
(sqlite3.OperationalError) near "or": syntax error
[SQL: SELECT Elementary, Middle, or High School AS School_Type, AVG(SAFETY_SCORE) AS Average_Safety_Score FROM CHICAGO_PUBLIC_SCHOOLS GROUP BY Elementary, Middle, or High School]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


### Problem 7

##### List 5 community areas with highest % of households below poverty line


In [37]:
%sql SELECT COMMUNITY_AREA_NAME FROM CENSUS_DATA ORDER BY PERCENT_HOUSEHOLDS_BELOW_POVERTY DESC LIMIT 5



 * sqlite:///FinalDB.db
Done.


[('Riverdale',),
 ('Fuller Park',),
 ('Englewood',),
 ('North Lawndale',),
 ('East Garfield Park',)]

### Problem 8

##### Which community area is most crime prone? Display the coumminty area number only.


In [39]:
%sql SELECT COMMUNITY_AREA_NUMBER, COUNT(*) AS INCIDENT_COUNT FROM CHICAGO_CRIME_DATA GROUP BY COMMUNITY_AREA_NUMBER ORDER BY INCIDENT_COUNT DESC LIMIT 1


 * sqlite:///FinalDB.db
Done.


[(25.0, 43)]


Double-click **here** for a hint

<!--
Query for the 'community area number' that has most number of incidents
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


In [42]:
%sql SELECT COMMUNITY_AREA_NAME FROM CENSUS_DATA WHERE HARDSHIP_INDEX IN (SELECT MAX(HARDSHIP_INDEX) FROM CENSUS_DATA);


 * sqlite:///FinalDB.db
Done.


[('Riverdale',)]

### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


In [46]:
%%sql
SELECT C.COMMUNITY_AREA_NAME
FROM CENSUS_DATA C
WHERE C.COMMUNITY_AREA_NUMBER = (
    SELECT CC.COMMUNITY_AREA_NUMBER
    FROM CHICAGO_CRIME_DATA CC
    GROUP BY CC.COMMUNITY_AREA_NUMBER
    ORDER BY COUNT(*) DESC
    LIMIT 1
);


 * sqlite:///FinalDB.db
Done.


[('Austin',)]

## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>
<h4>Abhishek Gagneja</h4>
<!--
## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
|2023-10-18  | 2.6     | Abhishek Gagneja  | Modified instruction set |
| 2022-03-04 | 2.5     | Lakshmi Holla     | Changed markdown.                   |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |
-->
## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
